### PR TITLE
feat: Implemented API key authentication mechanism

### PR DESCRIPTION
**Pull request by JigarJoshi**
## Describe your changes
Implemented API key authentication mechanism.
Implemented WhoAmI API.

## How best to test these changes
Added tests.

### CODE REVIEW

File: api/server/v1/marshaler.go [hunk 1]
```diff
@@ -1004,6 +1004,40 @@ func (x *ListInvoicesRequest) UnmarshalJSON(data []byte) error {
...
+	if err := jsoniter.Unmarshal(data, &mp); err != nil {
+		return err
+	}
+
+	for key, value := range mp {
+		var v any
+
+		switch key {
+		case "key_type":
+			var keyType string
```
Comment (efirs) [inline api/server/v1/marshaler.go:1020]: just `v = &x.KeyType` for project as well.

File: server/middleware/auth.go [hunk 2]
```diff
@@ -135,10 +136,10 @@ func GetJWTValidators(config *config.Config) []*validator.Validator {
 	return jwtValidators
 }
 
-func measuredAuthFunction(ctx context.Context, jwtValidators []*validator.Validator, config *config.Config, cache gcache.Cache) (context.Context, error) {
+func measuredAuthFunction(ctx context.Context, jwtValidators []*validator.Validator, config *config.Config, cache gcache.Cache, a auth.Provider) (context.Context, error) {
```
Comment (efirs) [inline server/middleware/auth.go:139]: You are becoming professional Go developer! :)

File: api/server/v1/marshaler.go (hunk 1 above)
Comment (JigarJoshi) [reply api/server/v1/marshaler.go:1020]: fixed it. Thanks

File: server/services/v1/auth/gotrue.go [hunk 3]
```diff
@@ -561,13 +617,18 @@ func _listAppKeys(ctx context.Context, g *gotrue, project string) ([]*appKeyInte
 			CreatedAt:   createdAtMillis,
 			Project:     appMetadata.Project,
 		}
+		if appMetadata.KeyType == "" || appMetadata.KeyType == AppKeyTypeCredentials {
+			appKey.KeyType = AppKeyTypeCredentials
+		} else if appMetadata.KeyType == AppKeyTypeApiKey {
+			appKey.KeyType = AppKeyTypeApiKey
```
[the remark this answers is not in the document]
Comment (JigarJoshi) [reply server/services/v1/auth/gotrue.go:623]: There is a validation in place that doesn't accept any other value.

[the remark this answers is not in the document]
Comment (JigarJoshi) [reply server/services/v1/auth/gotrue.go]: made updates

[the remark this answers is not in the document]
Comment (JigarJoshi) [reply server/services/v1/auth/gotrue.go]: made updates

File: server/services/v1/auth/gotrue.go [hunk 4]
```diff
@@ -185,8 +213,17 @@ func (g *gotrue) CreateGlobalAppKey(ctx context.Context, req *api.CreateGlobalAp
 	}, nil
 }
 
-func _updateAppKey(ctx context.Context, g *gotrue, id string, name string, description string) error {
-	email := fmt.Sprintf("%s%s", id, g.AuthConfig.Gotrue.UsernameSuffix)
+func _getEmail(clientId string, appKeyType string, g *gotrue) string {
+	if appKeyType == "" || appKeyType == AppKeyTypeCredentials {
```
Comment (efirs) [inline server/services/v1/auth/gotrue.go:217]: here too.
```
suffix := g.AuthConfig.Gotrue.UsernameSuffix
if appKeyType == AppKeyTypeApiKey {
    suffix = g.AuthConfig.ApiKeys.EmailSuffix
}

return fmt.Sprintf("%s%s", suffix)
```

File: server/services/v1/auth/gotrue.go [hunk 5]
```diff
@@ -654,6 +720,88 @@ func (g *gotrue) GetAccessToken(ctx context.Context, req *api.GetAccessTokenRequ
...
+	// parse JSON response
+	var getUserJsonMap map[string]any
+	err = json.Unmarshal(getUserResBody, &getUserJsonMap)
+	if err != nil {
+		log.Err(err).Msg("Failed to deserialize response into JSON")
+		return nil, errors.Internal("Failed to validate the api key")
+	}
+
+	// validate password
+	password := getUserJsonMap["encrypted_password"].(string)
```
[the remark this answers is not in the document]
Comment (JigarJoshi) [reply server/services/v1/auth/gotrue.go]: good point.

File: server/services/v1/auth/gotrue.go [hunk 6]
```diff
@@ -454,9 +501,15 @@ type appKeyInternal struct {
 	CreatedBy   string
 	CreatedAt   int64
 	Project     string
+	KeyType     string
 }
 
-func _listAppKeys(ctx context.Context, g *gotrue, project string) ([]*appKeyInternal, error) {
+func _listAppKeys(ctx context.Context, g *gotrue, project string, keyType string) ([]*appKeyInternal, error) {
+	if keyType != "" {
```
[the remark this answers is not in the document]
Comment (JigarJoshi) [reply server/services/v1/auth/gotrue.go]: simplified the other way that fits more naturally.

File: server/services/v1/auth/gotrue.go [hunk 7]
```diff
@@ -687,7 +834,7 @@ func getAccessTokenUsingClientCredentialsGotrue(ctx context.Context, clientId st
 	payloadValues := url.Values{}
 	payloadValues.Set("username", clientId)
 	payloadValues.Set("password", clientSecret)
-
+	log.Error().Interface("payload", payloadValues).Msg("Payload")
```
Comment (efirs) [inline server/services/v1/auth/gotrue.go:837]: Is it debug message?

File: server/services/v1/auth/gotrue.go [hunk 8]
```diff
@@ -550,7 +616,7 @@ func _listAppKeys(ctx context.Context, g *gotrue, project string) ([]*appKeyInte
 			// parse string time to millis using rfc3339 format
 			createdAtMillis = readDate(createdAtStr)
 		}
-
+		log.Info().Interface("metadata", appMetadata).Msg("AppMetadata")
```
Comment (efirs) [inline server/services/v1/auth/gotrue.go:619]: Is this debug message?

Comment (JigarJoshi) [reply server/services/v1/auth/gotrue.go]: yes it was left out by mistake. removed it.

File: server/services/v1/auth/gotrue.go (hunk 7 above)
Comment (JigarJoshi) [reply server/services/v1/auth/gotrue.go]: yes it was left out by mistake removed it.

**Review by efirs — APPROVED** [review]

**Pull request merged**